### ISSUE TITLE
Using openTSNE (with/without exaggeration) on your datasets

### ISSUE THREAD

**Issue by dkobak** [rozyangno/sce#5]
I ran openTSNE (https://github.com/pavlin-policar/openTSNE) on all your datasets and I thought you may find the following observations interesting:

1. t-SNE can run on all datasets, including HIGGS, in the default 750 iterations with all default parameters (openTSNE scales the learning rate with the sample size). It is also reasonably fast, e.g. ~50 min for HIGGS on my laptop.
2. t-SNE with late exaggeration ~4 produces results very similar to UMAP.
3. SCE roughly corresponds to t-SNE with much larger exaggeration than 4, however the exact value differs strongly between datasets (I don't know why!): for MNIST I get a good match with exaggeration ~4, for FLOWCYTOMETRY -- ~12, for IJCNN, SHUTTLE, HIGGS -- ~24, TOMORADAR -- ~50. Puzzling! 

I post below all my embeddings. Each figure shows t-SNE with exaggerations 1, 4, 12, 24, 50. I post my code below.
![MNIST](https://user-images.githubusercontent.com/8970231/195523693-ac54fe98-8505-4e43-87b3-866c0b59aa0d.png)
![IJCNN](https://user-images.githubusercontent.com/8970231/195523718-7cc1963c-4924-44b5-ad91-daf72d480c47.png)
![SHUTTLE](https://user-images.githubusercontent.com/8970231/195523733-f9f9a72b-9a85-40f5-8b66-5b1c5684040f.png)
![TOMORADAR](https://user-images.githubusercontent.com/8970231/195523752-173f704d-7a1f-4abe-9b86-058b9c306673.png)
![FLOWCYTOMETRY](https://user-images.githubusercontent.com/8970231/195523765-c972d2dd-7747-4561-bbc1-78ebd2f25666.png)
![HIGGS](https://user-images.githubusercontent.com/8970231/195523782-2000acbe-c1bc-4eac-9971-07668ec305a7.png)

And HIGGS figure plotted with smaller dot size, to emphasize some structure:
![HIGGS-001](https://user-images.githubusercontent.com/8970231/195523988-0489466e-7dbe-4740-80b7-b59f68cea541.png)

```python
import numpy as np
import pylab as plt
import seaborn as sns
from openTSNE import TSNE
from openTSNE.affinity import Affinities
import h5py
from scipy.sparse import csc_matrix

def load_data(dataset):
    f = h5py.File(f'P_{dataset}.mat', 'r')
    P_data = np.squeeze(f['P']['data'][()].T)
    P_ir   = np.squeeze(f['P']['ir'][()].T)
    P_jc   = np.squeeze(f['P']['jc'][()].T)
​
    P = csc_matrix((P_data, P_ir, P_jc)).tocsr()
    P /= np.sum(P)
    
    return P

def compute_tsnes(P, exaggerations = [1, 4, 12, 24, 50], verbose = False):
    Zs = []

    A = Affinities()
    A.P = P
​
    for ex in exaggerations:
        if not verbose:
            print('.', end='')
        if ex <= 12:
            Z = TSNE(n_jobs=-1, exaggeration=ex,
                     initialization='random', random_state=42,
                     verbose=verbose).fit(affinities = A)
        else:
            Z = TSNE(n_jobs=-1, exaggeration=ex, 
                     early_exaggeration=ex, learning_rate = P.shape[0]/ex,
                     initialization='random', random_state=42,
                     verbose=verbose).fit(affinities = A)
    
        Zs.append(Z)
​
    if not verbose:
        print('')
​
    return Zs

def make_figure(Zs, title=None, figsize=(9, 2.2), s=.1):
    plt.figure(figsize=figsize)
​
    for i,Z in enumerate(Zs):
        plt.subplot(1, len(Zs), i+1)
        plt.axis('equal')
        plt.scatter(Z[:,0], Z[:,1], s=s, ec='none')
        plt.xticks([])
        plt.yticks([])
​
    sns.despine(left=True, bottom=True)
    if title:
        plt.suptitle(title)
    plt.tight_layout()

datasets = ['MNIST', 'SHUTTLE', 'IJCNN', 'TOMORADAR', 'FLOWCYTOMETRY', 'HIGGS']
​
for dataset in datasets:
    P = load_data(dataset)
    Zs = compute_tsnes(P, verbose=True)
    make_figure(Zs, title=dataset)
    plt.savefig(f'{dataset}.png', dpi=200)
```

**Comment by rozyangno** [comment]
Please notice two things:
1) Early exaggerations are used only in t-SNE initialization. To get the final t-SNE result, you must use exaggeration=1.
2) Even if you use large exaggerations throughout the iterations, the amount of exaggerations differs a lot across the six datasets. In contrast, SCE uses alpha=0.5 for all datasets, which is consistent and much more accessible.

**Comment by dkobak** [comment]
Yes I agree.

Anyway, this is not really an "issue", so feel free to close it. I just thought you might find these results interesting.

**Comment by rozyangno** [comment]
Thanks. Now I close it.
